### PR TITLE
Bluetooth: host: Fix packet reording in ATT under high load

### PR DESCRIPTION
**Pull request by joerchan**
Fix packet reordering in ATT when putting the host under high load.
In certain conditions a packet can be transmitted while there is
elements in the TX queue in bt_att_send. This means that the packet will
skip ahead in the TX queue leading to packet re-ordering.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/host/att.c [hunk 1]
```diff
@@ -436,6 +436,31 @@ static int bt_att_chan_send(struct bt_att_chan *chan, struct net_buf *buf,
...
+		err = bt_att_chan_send(chan, buf, NULL);
+		if (err >= 0) {
+			break;
+		}
+	}
+
+	if (err < 0) {
+		/* Push it back if it could not be send */
+		k_queue_prepend(&att->tx_queue._queue, buf);
+			return;
```
Comment (jhedberg) [inline subsys/bluetooth/host/att.c:460]: Seems like an unneeded (and also incorrectly indented) line

**Review by Thalley — APPROVED** [review]
Besides https://github.com/zephyrproject-rtos/zephyr/pull/33775/files#r603294703 this LGTM. 

If I understand the change correctly, it is simply changing the queuing from append to prepend in the `att_send_process` case? 

Does this solve the issue with EATT you mentioned before?

**Comment by joerchan** [comment]
> Besides https://github.com/zephyrproject-rtos/zephyr/pull/33775/files#r603294703 this LGTM.
> 
> If I understand the change correctly, it is simply changing the queuing from append to prepend in the `att_send_process` case?
> 
> Does this solve the issue with EATT you mentioned before?

I have only tested this in non-EATT context. I can't yet say for sure that EATT has a problem with reordering or not with this change.

The change is to always go through the queue before sending. So always append it to the queue, and then try to send the head of the queue, if that fails then put it back by prepending it to the queue.

**Comment by joerchan** [comment]
@Vudentz I think that while this fix might fix packet reordering for EATT it could potentially impact the performance, since it is all now going through the queue instead of going over a channel immediately.

To me it seems like it would be wiser to expose this to the application so that we could give services their own dedicated channels for the most important services, audio services for example.
Packet reordering between different services wouldn't be an issue, but within a single service it presents a problem.
For example DFU needs to transmit all of the notifications in order.

**Comment by Vudentz** [comment]
> @Vudentz I think that while this fix might fix packet reordering for EATT it could potentially impact the performance, since it is all now going through the queue instead of going over a channel immediately.
> 
> To me it seems like it would be wiser to expose this to the application so that we could give services their own dedicated channels for the most important services, audio services for example.
> Packet reordering between different services wouldn't be an issue, but within a single service it presents a problem.
> For example DFU needs to transmit all of the notifications in order.

Well in case of DFU with EATT I would argue that can use a dedicated L2CAP channel directly given that both peer should support it (for EATT), as for exposing bt_att_chan that would need a whole set of new APIs, not sure if it worth the trouble given that all the heavy traffic use cases (aka GATT attribute used as serial port) could be changed to use dedicated L2CAP channel directly instead of reserving EATT channels.

**Review by Vudentz — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**